pkg/network/opencontrail/cni/cni/cni.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2016 Juniper Networks, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/* Common CNI plugin for Kubernetes and Mesos. */
+package cni
+
+import (
+	"../args"
+	"../ipc"
+	"fmt"
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/vishvananda/netlink"
+	"log"
+	"net"
+)
+
+const CniVersion = "0.2.0"
+
+/****************************************************************************
+ * ADD message handlers
+ ****************************************************************************/
+// Create the veth pairs and get host-os name for interface
+func processContainerAdd(netns ns.NetNS, cniArgs *args.CniArgs, mac string,
+	typesResult *types.Result) (string, error) {
+	var hostIfName string
+	err := netns.Do(func(hostNS ns.NetNS) error {
+		// create veth pair in container and move host end into host netns
+		link, _, err := ip.SetupVeth(cniArgs.IfName, cniArgs.ContrailArgs.Mtu,
+			hostNS)
+		if err != nil {
+			return err
+		}
+
+		hostIfName = link.Attrs().Name
+
+		err = ipam.ConfigureIface(cniArgs.IfName, typesResult)
+		if err != nil {
+			return err
+		}
+
+		hwAddr, mac_err := net.ParseMAC(mac)
+		if mac_err != nil {
+			log.Println("Error parsing MAC address ", mac, " Error ", mac_err)
+			return mac_err
```

Comment:
This return is part of anonymous function passed to netns.Do(). The anonymous function returns one value only